### ISSUE TITLE
sk_pref validation user journey

### ISSUE THREAD

**Issue by istvanp1** [skpref/skpref#9]
This issue is a discussion around how users will be able to validate their models using the task structure.
Currently this is how a user can validate using a train and test set:
```python
from tbc_folder import BradleyTerry
from task import ChoiceTask
import pandas as pd

# Split my train and test sets
NBA_results = pd.read_csv(NBA_file_loc)
season_split = 2016
train_data = NBA_results[NBA_results.season_start == season_split].copy()
test_data = NBA_results[NBA_results.season_start == season_split+1].copy()

# Create a task for the different sets
NBA_results_task_train = ChoiceTask(primary_table=train_data, primary_table_alternatives_names=['team1', 'team2'], primary_table_target_name ='team1_wins',
                                    target_column_correspondence='team1', features_to_use=None)
NBA_results_task_predict = ChoiceTask(primary_table=test_data, primary_table_alternatives_names=['team1', 'team2'], primary_table_target_name ='team1_wins',
                                    target_column_correspondence='team1', features_to_use=None)

# fit my model
mybt = BradleyTerry(method='BFGS', alpha=1e-5)
mybt.fit_task(NBA_results_task_train)

# predict my model
mybt.predict_task(NBA_results_task_test)
```

Though I admit this sometimes might be the necessary way of doing it, we might be able to create some extra functionality where the user doesn't has to copy paste the ChoiceTask with only one difference being the primary_table name.

Suggestion 1:
Allow train and test sets to be passed into ChoiceTask which by default predicts the test_set in this case, but can be set to False.
```python
NBA_results_task = ChoiceTask( primary_table_alternatives_names=['team1', 'team2'], primary_table_target_name ='team1_wins',
                                    target_column_correspondence='team1', features_to_use=None, training_table = train_data, test_table = test_data)

# fit my model
mybt = BradleyTerry(method='BFGS', alpha=1e-5)
mybt.fit_task(NBA_results_task)

# predict my model
mybt.predict_task(NBA_results_task, predict_test=True)
```

Suggestion 2:
fit task assumes that the primary_table is being predicted, but it can be overwritten with another table, for which the same data structure is being assumed.
```python
NBA_results_task = ChoiceTask(primary_table=train_data, primary_table_alternatives_names=['team1', 'team2'], primary_table_target_name ='team1_wins',
                                    target_column_correspondence='team1', features_to_use=None)

# fit my model
mybt = BradleyTerry(method='BFGS', alpha=1e-5)
mybt.fit_task(NBA_results_task)

# predict my model
mybt.predict_task(NBA_results_task, new_df=test_data)
```

**Comment by timokau** [comment]
Why not separate construction and instantiation?

```python
from tbc_folder import BradleyTerry
from task import ChoiceTask
import pandas as pd

# Split my train and test sets
NBA_results = pd.read_csv(NBA_file_loc)
season_split = 2016
train_data = NBA_results[NBA_results.season_start == season_split].copy()
test_data = NBA_results[NBA_results.season_start == season_split + 1].copy()

# Create a task
task = ChoiceTask(
    primary_table=train_data,
    primary_table_alternatives_names=["team1", "team2"],
    primary_table_target_name="team1_wins",
    target_column_correspondence="team1",
    features_to_use=None,
)

# Load data
NBA_results_task_train = task.load(train_data)
NBA_results_task_predict = task.load(test_data)

# fit my model
mybt = BradleyTerry(method="BFGS", alpha=1e-5)
mybt.fit_task(NBA_results_task_train)

# predict my model
mybt.predict_task(NBA_results_task_test)
```

**Comment by istvanp1** [comment]
That could also work, but then syntactically its confusing to define train_data as the primary_table of the task and then having to load train_data into the task. Perhaps along these lines would be better:
```python
# Create a task
NBA_results_task_train= ChoiceTask(
    primary_table=train_data,
    primary_table_alternatives_names=["team1", "team2"],
    primary_table_target_name="team1_wins",
    target_column_correspondence="team1",
    features_to_use=None,
)

# fit my model
mybt = BradleyTerry(method="BFGS", alpha=1e-5)
mybt.fit_task(NBA_results_task_train)

# Create a copy of the task but with a new primary table
NBA_results_task_predict = NBA_results_task_train.update_primary_table(test_data)

# predict my model
mybt.predict_task(NBA_results_task_test)
```
Is it possible though in python to return a new copy of the instance of the class by a method of the class? It could possible to overwrite the the existing NBA_results_task_train object, but that probably wouldn't be desirable.

**Comment by istvanp1** [comment]
Actually one more point of discussion would be potentially beneficial here. Markus has pointed out to me the following opportunity:
```python
# classes
class Estimator():    
    def fit(self, task, data):
        task.check_data_compatibility(data)        
        view = data.create_view(task)        
        # do something with view
        return self
class Task():    
    def check_data_compatibility(self, data):
        # run checks if columns are present, etc
# workflow
task = Task()
data = load_data()
train, test = train_test_split(data)
estimator = Estimator()
estimator.fit(train, task)
estimator.predict(test)
```
The benefits of this method is that users don't have to create different tasks for what intuitively feels like is the task. Its a significant step away from the original idea of loading primary tables into the task function. I'm not convinced at the moment that this would be necessarily a bad thing, what do you think?

**Comment by mloning** [comment]
The line `task.check_data_compatibility(data)` above should probably be inside of the `create_view` method.

**Comment by prithagupta** [comment]
The problem with solution I think is that we cannot use the split for the data without knowing for which task, we are using that data for. 

For example in the case of sushi, we have 5000 rankings are on 10 sushi options provided to them and the other 5000 rankings are on top-10 sushis out of 100 ones. SO the train-test-split could be done in multiple ways, we can learn on one set and predict on others, or we can combine all the rankings and then split them. 

there are some datasets, where the split is provided.

I have another question, does Task mean it represents any task, OrderedTask, ChoiceTask etc?

**Comment by fkiraly** [comment]
Ah, the good old "what about train/test splits vs tasks" discussion, about which I'm not sure whether I have a fully conclusive answer. Relevant options off the top of my head:
(i) train/test split(s), as in CV object, an optional field of task. If given, task is interpreted as the task with these specific splits for evaluation.
(ii) splitting or sub-indexing can be applied to tasks.
(iii) the task is separate from the data entirely, i.e., retains only the relational and statistical data model, which will be the same for train and test set.
(iv) the task contains the data, and validation is done via a separate "benchmarking" class or function. To this, intsructions on the splitting are passed separately.

**Comment by fkiraly** [comment]
Just to make sure: I think nothing in the discussion is specific to choice models, except perhaps the secondary issue of how train/test splits would apply to secondary tables.

**Comment by istvanp1** [comment]
I'm not quite sure I'm on the same page at the moment with this being only specific for choice models, train/test splits would be done for any type of model right?

**Comment by fkiraly** [comment]
What I mean, the API discussion: all this discussion could be applied to "tasks tor supervised learning" verbatim. Yes or no?
The one difference being that "tasks vs no tasks" *is* somewhat specific, as in that "no tasks" leads to cumbersome and repetitive method signatures. True?
(trying to cross-check my understanding with yours)

**Comment by timokau** [comment]
(iii)

> That could also work, but then syntactically its confusing to define train_data as the primary_table of the task and then having to load train_data into the task.

I made a mistake in my example, I didn't mean to also include the `primary_table` in the task generation. So the task genration would be totally agnostic to the training/test data. I think that is also @fkiraly's optikon (iii).

> Is it possible though in python to return a new copy of the instance of the class by a method of the class?

That should certainly be possible.

**Comment by istvanp1** [comment]
@fkiraly, yes this is related for tasks for supervised learning. There is definitely a benefit to having tasks define the data structure, and not having it will lead to confusingly much having to be specified in the fit method.
I think what we are experiencing in this discussion is the phenomenon that there are many good solutions in coding (also many bad ones) and it can be tough to make a choice in some cases.

What might help is if we write down a collectively agreed principles of coding and some way to assess how proposed code fits to those principles. I will open a new ticket called "The skpref coding principles" and tag everyone I can. For those that I can't please also contribute your opinion. I know this is a bit of a detour, but I believe if we get this right it can speed up a few decisions in the future and also can part of a valuable guide to any future contributors. Also the Christmas wind-down is probably a good time to have these chats.

Once we have settled our principles of coding we can come back here and assess how each of these options fits with that.

**Comment by istvanp1** [comment]
Also @timokau cool, then suggestion (iii) by @fkiraly , and the suggestions by @mloning are actually along the same lines.

**Comment by istvanp1** [comment]
you can see the discussion on design principles in issue #16
